### ISSUE TITLE
Feature Request: allow change file extension of generated files from `.ts` 

### ISSUE THREAD

**Issue by bluelovers** [microsoft/TypeScript#49462]
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

add something like 

```
{
  "compilerOptions": {
    "module": "nodenext",
    "targetExtension": ".cjs",
}
```

## 📃 Motivating Example

- when `targetExtension` is `.cjs` , all `.ts` will emit as `.cjs`, but `.mts` still is emit as `.mjs`
- when `targetExtension` is `.mjs` , all `.ts` will emit as `.mjs`, but `.cts` still is emit as `.cjs`

**Comment by IllusionMH** [comment]
What are exact use cases and what problem it should resolve?
From first look TS would also need to rewrite extension in generated imports that in no go at this moment #49083 (if I haven't missed anything).

**Comment by Josh-Cena** [comment]
Major use-case is building dual-package with `tsc` without any postbuild script, I think.

**Comment by milesj** [comment]
You shouldn't be dual building packages anyways (use a wrapper), so I prefer the `.cts`/`.mts` constraint that TS indirectly enforces.

**Comment by bluelovers** [comment]
i don't wanna make files of .cts and .mts they are same context
so i wanna one .ts can be .cjs and .mjs

**Comment by azu** [comment]
I've met same issue when I building dual package.

e.g. https://github.com/azu/check-ends-with-period/tree/v2.0.0 (It is invalid example as dual package)
TypeScript source code is in`src/*.ts` and `package.json` has `"type": "module"` field.
Also, this pacakge has two tsconfig files.

- [`tsconfig.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.json) : generates esm to `module/*.js`
- [`tsconfig.cjs.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.cjs.json): generates cjs to `lib/*.js`

I've defined `exports` field as follows, but this package is treated as ESM because `"type": "module"` is defined. 
As a result, This package can not be requred from CJS.(without dynamic import)

```
  "exports": {
    ".": {
      "require": "./lib/check-ends-with-period.js",
      "import": "./module/check-ends-with-period.js"
    }
  },
```

If `targetExtension` option exists, I can resolve this issue by tsconfig file.

**Comment by milesj** [comment]
Might I suggest packemon: https://packemon.dev/

Also, why exactly are you dual building? You run the risk of the dual package hazard: https://nodejs.org/api/packages.html#dual-package-hazard It's better to use an ES module wrapper.

**Comment by bluelovers** [comment]
> 

i think this only work on nodejs
does browser support it?

```
// ./node_modules/pkg/wrapper.mjs
import cjsModule from './index.cjs';
export const name = cjsModule.name;
```

**Comment by milesj** [comment]
Yes of course, but not if you're using .mjs. At least in @azu's example, their ESM code should be shipped to the browser with .js, and CJS code to Node.js with .cjs (or even just .js too). 

We also just need more information, as we're making many assumptions here. The original post doesn't contain much.

**Comment by Josh-Cena** [comment]
Wouldn't we be emitting `.js` (+ type: module) + `.cjs` when building dual-package purely for Node anyway? It's usually unnecessary to have explicit extensions for both sets of module types. Also, browsers can handle .mjs as well, as long as the MIME type is JavaScript. But to author dual-package of _any kind_, whether targeting Node or browser, we need at least one subset of the output to have an extension different from the other, and that would require TS to be configurable about this. But I agree we lack some context here.

**Comment by alvis** [comment]
I'm pro @bluelovers's suggestion. The use case for me is publishing dual ESM/CommonJS packages for nodejs.

There are many reasons why we need to publish dual ESM/CommonJS packages. For instance, not until 4.7, TS doesn't even allow a node application in commonjs to consume a pure ESM library due to the lack of support on `await import(...)`. So I really find it annoying some package authors publish pure ESM packages.

**Comment by cefn** [comment]
@milesj the use of an ESM wrapper around CommonJS defeats tree-shaking, doesn't it? Given ESM has broken the whole ecosystem to try and achieve results like tree-shaking, having the recommended way to align with ESM being to throw away its core features is disappointing. Correct me if there is a reasonable way to get both. My expectation is to follow the principle that modules should be stateless - a good practice I don't ever find the need to violate. Then I understand there are no concerns with dual building.

**Comment by milesj** [comment]
If you have a dual package, and some other package in CJS context requires your package, and another package in ESM context imports your package, you'll end up with 2 copies of your package. For node this doesn't matter too much unless there's some kind of global/shared state, but for bundlers this is bad.

**Comment by gfortaine** [comment]
> I've met same issue when I building dual package.
> 
> e.g. https://github.com/azu/check-ends-with-period/tree/v2.0.0 (It is invalid example as dual package) TypeScript source code is in`src/*.ts` and `package.json` has `"type": "module"` field. Also, this repository has two tsconfig files.
> 
> * [`tsconfig.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.json) : generates esm to `module/*.js` from `src/*.ts`
> * [`tsconfig.cjs.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.cjs.json): generates cjs to `lib/*.js` from `src/*.ts`
> 
> I've defined `exports` field as follows, but this package was treated as ESM because `"type": "module"` is defined. As a result, This package can not be requred from CJS without dynamic import. (Node.js treats `*.js` file as ESM by `"type": "module"`)
> 
> ```
>   "exports": {
>     ".": {
>       "require": "./lib/check-ends-with-period.js",
>       "import": "./module/check-ends-with-period.js"
>     }
>   },
> ```
> 
> I could not found just works solution without using bundler/post scripts.
> 
> If `targetExtension` option exists, I can resolve this issue by tsconfig file.
> 
> * [`tsconfig.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.json) + `"targetExtension": ".mjs"`,: generates esm to `module/*.mjs` from `src/*.ts`
> * [`tsconfig.cjs.json`](https://github.com/azu/check-ends-with-period/blob/v2.0.0/tsconfig.cjs.json) + `"targetExtension": ".cjs"`,: generates cjs to `lib/*.cjs` from `src/*.ts`
> 
> However, this option will need to rewrite import path of source code. It oppsite [TypeScript's design goal](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals).

@azu @milesj It looks like that [tsc-multi](https://github.com/tommy351/tsc-multi) might be worth exploring cc @tommy351 https://github.com/microsoft/TypeScript/issues/18442#issuecomment-1369110873

**Comment by azu** [comment]
I've created [tsconfig-to-dual-package](https://github.com/azu/tsconfig-to-dual-package) for avoiding my issue that is described in https://github.com/microsoft/TypeScript/issues/49462#issuecomment-1152860276
This tool add package.json which is `{ "type": "module" }` or `{ "type": "commonjs" }` based on tsconfig's `module` and `outDir` options.
It resolve my issue by adding each type `pacakge.json` to `lib/`(CJS) and `module/`(ESM) instead of use `.cjs` and `.mjs`
This behavior is described in following:

- https://github.com/nodejs/node/issues/34515#issuecomment-664209714
- https://nodejs.org/api/packages.html#type

In other words, Both(CJS and ESM) are `*.js`.
The result is much closer to what I wanted to do.

```bash
$ tsc -p . && tsc -p ./tsconfig.cjs.json && tsconfig-to-dual-package
#                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#                                          \ I want to remove it!
```

This approch pros is that no require addtional build/transpiler tool(no modify output code of `tsc`).
Cons is that need to copy `package.json` to outDir.
(Of course, dual package hazard is in here, but this hazard is also in browser's iframe/realm or multiple versions of the same library. Not ideal, I believe the actual harm is limited.)


📝 Note

TypeScript may change the moduleResolution, but did not likely change the output.

- https://github.com/microsoft/TypeScript/issues/50152

I feel like there is [confusion](https://github.com/npm/rfcs/issues/665) [everywhere](https://github.com/nodejs/node/issues/46006) about ESM support. Therefore, I thought that this is not an issue of TypeScript configuration, but rather an ecosystem-wide issue that needs to move forward.

**Comment by gfortaine** [comment]
> I've created [tsconfig-to-dual-package](https://github.com/azu/tsconfig-to-dual-package) for avoiding my issue that is described in [#49462 (comment)](https://github.com/microsoft/TypeScript/issues/49462#issuecomment-1152860276) This tool add package.json which is `{ "type": "module" }` or `{ "type": "commonjs" }` based on tsconfig's `module` and `outDir` options. It resolve my issue by adding each type `pacakge.json` to `lib/`(CJS) and `module/`(ESM) instead of use `.cjs` and `.mjs` This behavior is described in following:
> 
> * [Improve documentation on Dual Module Packages nodejs/node#34515 (comment)](https://github.com/nodejs/node/issues/34515#issuecomment-664209714)
> * https://nodejs.org/api/packages.html#type
> 
> In other words, Both(CJS and ESM) are `*.js`. The result is much closer to what I wanted to do.
> 
> ```shell
> $ tsc -p . && tsc -p ./tsconfig.cjs.json && tsconfig-to-dual-package
> #                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> #                                          \ I want to remove it!
> ```
> 
> This approch pros is that no require addtional build/transpiler tool(no modify output code of `tsc`). Cons is that need to copy `package.json` to outDir. (Of course, dual package hazard is in here, but this hazard is also in browser's iframe/realm or multiple versions of the same library. Not ideal, I believe the actual harm is limited.)
> 
> 📝 Note
> 
> TypeScript may change the moduleResolution, but did not likely change the output.
> 
> * [A Proposal For Module Resolution #50152](https://github.com/microsoft/TypeScript/issues/50152)
> 
> I feel like there is [confusion](https://github.com/npm/rfcs/issues/665) [everywhere](https://github.com/nodejs/node/issues/46006) about ESM support. Therefore, I thought that this is not an issue of TypeScript configuration, but rather an ecosystem-wide issue that needs to move forward.

@mobsense @jwalton @owenallenaz

**Comment by fabis94** [comment]
I don't get the "you shouldn't be dual building" opinion, clearly there's a lot of fragmentation in the ecosystem currently and while everyone's moving in the direction of using ESM, there's still a lot of CJS projects and that's not going to change overnight. Because of this reason if you're building a library, it's important to output both in CJS and MJS.

Additionally, `"type": "module"` and `"moduleResolution": "nodenext"/"node16"` are the future, and in a project with these set any CJS build results outputted with a .js extension simply won't work.

Thus, it makes perfect sense to me that instead of outputting results with a vague ".js" extension which causes issues and headaches, you'd want to have explicit .cjs and .mjs extensions for CommonJS and ESModule modules respectively. On top of that, outputted declaration files should have d.cjs and .d.mts extensions respectively, cause another issue I've ran into multiple times recently is library authors outputting results in .mjs and .cjs, but leaving the declaration files with the .d.ts extension which TypeScript then won't be able to pick up on.

**Comment by zhangyu1818** [comment]
I'm working on a CLI tool that's bundled with tsc. I set up `{"bin":"./dist/index.js"}` in package.json, but the command doesn't run in projects with `"type":"module"`. I noticed that `Prettier` is set to `"./bin/prettier.cjs"`. It works fine maybe because of its`.cjs` extension.

**Comment by leeren** [comment]
What are people currently using to work around this? Separate build steps with manual file extension rewriting?

**Comment by azu** [comment]
I believe that Dual Package can be achieved broadly as follows.

1. Write two source codes CJS and ESM respectively
    - Write two source codes by hand
    - Handwritten, which is expensive to maintain.
2. Generate the CJS code from one source code and make the ESM a wrapper that only imports CJS.
    - The majority of the source code is in CJS format.
    - The ESM becomes a wrapper that imports CJS.
    - e.g. packemon's [Automatic .mjs wrappers for .cjs inputs](https://packemon.dev/docs/features#automatic-mjs-wrappers-for-cjs-inputs)
3. Generate the ESM code from one source code and make the CJS a wrapper that only imports ESM.
    - The majority of the source code is in ESM format.
    - The reverse pattern of 2.
    - The CJS entry point is used to Dynamic Import the ESM.
    - Limitations: synchronous APIs cannot be provided from the CJS format.
    - e.g. [Prettier v3](https://github.com/prettier/prettier/blob/79d8a3983842d1d8a077f177ed8b983c1c596088/src/index.cjs), [Vite](https://github.com/vitejs/vite/pull/8178)
4. Generate CJS and ESM code from one source code
    - Some kind of build tool or script is required
    - What tools are needed will depend on the library
    - e.g.
      - [Packemon](https://packemon.dev/) uses [babel-plugin-cjs-esm-interop](https://www.npmjs.com/package/babel-plugin-cjs-esm-interop) for building
      - [rollup](https://rollupjs.org/) supports multiple outputs
          - [Marked](https://github.com/markedjs/marked) uses [rollup](https://github.com/markedjs/marked/blob/9a3d0892f8673b3e5d8aae7f94c1fca263700a39/rollup.config.js)
          - [Microbundle](https://github.com/developit/microbundle) is a wrapper of rollup
      - [tsconfig-to-dual-package](https://github.com/azu/tsconfig-to-dual-package) uses [package.json hacks](https://github.com/azu/tsconfig-to-dual-package#how-it-works). [Glob](https://github.com/isaacs/node-glob) does [same thing](https://github.com/isaacs/node-glob/blob/main/fixup.sh)

**Comment by morganney** [comment]
I stumbled upon this issue recently while wanting to build a dual package using babel to transpile and tsc to type check and generate declaration files. This file extension proliferation in the JS ecosystem is needless to say frustrating.

I just started a project [`babel-dual-package`](https://github.com/morganney/babel-dual-package) that takes `"type": "module"` packages and creates separate ESM and CJS builds with extensions correctly updated in `import`/`export` statements, `import()` expressions and filenames.  This includes declaration `.d.ts` files as well.  All you need to do is add your `exports` field in package.json to match the build output. If you use babel and typescript together it might be helpful.

As I mentioned, I just started (published yesterday) and have only used it on a node (i.e. not browser) project, so the feature list is a bare minimum at this point. It also requires a node version >= 20.1.0 atm. Any help or feedback is welcomed.

Example:

**tsconfig.json**
```json
{
  "compilerOptions": {
    "target": "es6",
    "module": "esnext",
    "moduleResolution": "nodenext",
    "declaration": true,
    "emitDeclarationOnly": true,
    "declarationDir": "dist",
    "isolatedModules": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  },
  "include": ["src"]
}
```

**tsconfig.cjs.json**

So `tsc` generates declaration files inside of a separate directory meant for the CJS build. I suppose `babel-dual-package` could just copy the ones from the ESM build and bypass this step.
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "CommonJS",
    "moduleResolution": "Node",
    "declarationDir": "dist/cjs"
  },
  "include": ["src"]
}
```

**babel.config.json**
```json
{
  "presets": [
    ["@babel/preset-env", {
      "modules": false
    }],
    "@babel/preset-typescript"
  ]
}
```

**package.json** (only the relevant bits)
```json
{
  "type": "module",
  "scripts": {
    "check-types": "tsc --noEmit",
    "build:types": "tsc && tsc -p ./tsconfig.cjs.json",
    "build:dual": "babel-dual-package --out-dir dist --extensions .ts,.mts,.cts src",
    "build": "npm run build:types && npm run build:dual"
  }
  "devDependencies": {
    "@babel/preset-env": "^7.22.5",
    "@babel/preset-typescript": "^7.22.5",
    "babel-dual-package": "^1.0.0-alpha.0",
    "typescript": "5.0.4"
  }
}
```

The gist is `babel-dual-package --out-dir dist --extensions .ts,.mts,.cts src`. You can `./node_modules/.bin/babel-dual-package --help` to see some options for configuring the output directory, supported file extensions, etc.

This will get you an ESM build in `dist` and a CJS build in `dist/cjs`. Define your `exports` accordingly.

**Comment by owenallenaz** [comment]
What we ended up doing is simply... not supporting ESM. In my eyes, the ESM ecosystem is simply not robust enough and we were spending more time trying to fight through it than actually solving real problems. Maybe once the tooling is there, but it's simply not. I want to be able to import using barrels like `import Foo from "./Foo"` where that references a file that is /foo/index.ts or /foo/index.tsx. It works in typescript natively, but when we compile for ESM it's busted. If developers want us to use ESM it needs to be ESM and no mandate that your entire toolset be built in ESM. If TS only worked with TS, it never would have got off the ground.

**Comment by WoodyWoodsta** [comment]
To add my two cents, regardless of whether you want to emit dual variants, `module: "commonjs"` should output files with either a `cjs` extension, or a `js` extensions.

At the moment, if you write your source code as `mts`, the compiler emits commonjs code but in `mjs` files which to me seems incorrect and broken. I don't think there should be a question of "why would you want to do that" - the compiler offers a `module` option, and so the options should output spec-compliant code, or should produce an error if the inputs don't comply with the config or output format.

**Comment by knightedcodemonkey** [comment]
If you want to update your ESM [specifiers](https://nodejs.org/api/esm.html#import-specifiers) pre/post build, check out [@knighted/specifiers](https://github.com/knightedcodemonkey/specifier). It will parse a file and update specifiers using a provided callback or regex map. Then write the updated source to disk using whatever file extension you want.

**Comment by knightedcodemonkey** [comment]
The fact that `.mts` files are **always** converted to the CJS module system anytime `--module commonjs` is used (despite the `--moduleResolution` used, or the `type` defined in package.json) **needs to be addressed first**. This breaks things, and makes building dual packages with `tsc` exclusively nearly impossible (ok, probably strictly impossible). This is clearly antithetical to [how Node determines module systems](https://nodejs.org/api/packages.html#determining-module-system).

See microsoft/TypeScript#54573 and all the related issues mentioned in this [comment](https://github.com/microsoft/TypeScript/issues/54573#issuecomment-1585134598).

The file extension nonsense is annoying, but that's how we let the two module systems that Node currently supports coexist.

**Comment by ceztko** [comment]
Today I also had the need for compiling `.ts` files to javascript files with a different extension. Instead of the proposed `"targetExtension": ".cjs"`, or `"targetExtension": ".mjs"` I suggest something like `"ensureQualifiedExtension" : "true"` that behaves like this:

- If we have `"module": "commonjs"` then the extension is always `.cjs`;
- If we have `"module": "es6"`, or `es2015`, `es2020`, `es2022`, `esnext` then the extension is always `.mjs`;
- If we have `"module": "node16"` or `nodenext` then the extension respects the rules for module detection as [documented](https://www.typescriptlang.org/docs/handbook/modules/reference.html#the-module-compiler-option) and the final extension can be either `.cjs` or `.mjs`.

**Comment by ruojianll** [comment]
I need this nowadays, but we may not need it in the future, so keep it silence.

**Comment by Tofandel** [comment]
> If we have "module": "commonjs" then the extension is always .cjs;
> If we have "module": "es6", or es2015, es2020, es2022, esnext then the extension is always .mjs;

I would tweak this a bit so that you get cjs in the first case if your package.json contains `"type": "module"` and `.js` otherwise
And vice versa for the second case, you get `.mjs` if package.json doesn't contain `"type": "module"` and `.js` otherwise

**Comment by ceztko** [comment]
> I would tweak this a bit so that you get .cjs in the first case if your package.json contains "type": "module" and .js otherwise
And vice versa for the second case, you get .mjs if package.json doesn't contain "type": "module" and .js otherwise

I think `"type": "module"` is evaluated only with `"module": "node16"` or `nodenext` automatic module detection. In my suggestion, the extension is driven first by the `module` setting in `tsconfig.json`, and automatically enforces a qualified extension which in some cases is really desired/required.

In my understanding your tweak, as you call it, would introduce a slack rule unconditionally evaluating `"type": "module"` unregarding of `module` setting in `tsconfig.json`, and would not supply a stable extension for people that need an early enforcement of the module convention.
 
Anyway, other than few thumbs up, I don't see any reaction that suggests the general experience will be improved any time soon.

**Comment by jedwards1211** [comment]
@RyanCavanaugh because many people have the default `"moduleResolution"` setting, which ignores export maps, the only way we can support stuff like `import * from 'foo/bar'` without any TS users running into problems is to have `bar.js` and `bar.mjs` files in the root of the package.

I'm pretty sure @azu's `tsconfig-to-dual-package` won't work for people with default `"moduleResolution"` because it builds the CJS and ESM into subdirectories: https://github.com/azu/tsconfig-to-dual-package/issues/15

I haven't seen any TS dual-package authoring tools that seem unproblematic.  The popular `tsc-alias` uses regexes to rewrite import statements, and I'm guessing breaks source maps too.

If `tsc` would just add quality ways to build dual packages, it would be a lot easier for everyone to avoid the subtle pitfalls of popular tools and approaches.

**Comment by RyanCavanaugh** [comment]
> I haven't seen any TS dual-package authoring tools that seem unproblematic

The whole reason we don't want to do dual-emit is that there doesn't seem to *be* an unproblematic way to do it. What can TS do that they can't?

**Comment by morganney** [comment]
If you set `moduleResolution` to `nodenext` it’s possible to generate dual builds with tsc using only one tsconfig.json and package.json file. Anybody authoring a library today that wants to support dual build should be using `nodenext` anyways.

**Comment by jedwards1211** [comment]
@RyanCavanaugh what makes general-purpose tools for dual-emit problematic is they have to either copy the source code to a temporary directory and codemod the import path extensions before compiling, running the risk of something breaking in the temporary location, or they have to destructively modify the source code in place before compiling.

Only `tsc` or `babel` (or maybe a 3rd-party tool that wraps the `tsc` internal API?) can codemod import paths in memory, transpile, and then output to a file with the desired extension.

I use a custom `babel` plugin for it, but of course most TS devs will not prefer this.

And even if I could make a 3rd-party `tsc` wrapper I could only make a fraction of TS devs aware of it compared to if it was a builtin `tsc` option.

Modifying the import path extensions is really simple compared to most of the transpilation that `tsc` does, changing the output file extensions is even simpler, and these options would be useful for a variety of dual package layouts.

@morganney are you talking about having both `.(c)ts` and `.(m)ts` in your source code?  I've dealt with this problem for years and the only dual package layout I've found that supports the widest variety of situations (subpath imports from the package, package users who aren't using `nodenext`, not simply re-exporting the CJS modules from ES wrappers) is to transpile `.ts` source to both CJS and ESM files side by side, provide an export map for people whose tools can consume it, and fall back to relative path resolution to either the CJS or ESM files for users whose tools ignore the export map.

**Comment by morganney** [comment]
@jedwards1211 

No, I’m talking about a build tool that runs tsc twice while updating the type in package.json and rewriting specifiers. Use whatever file extension you want.

**Comment by heath-freenome** [comment]
I just encountered a situtation where I want to make the output of a `commonjs` compile be `.cjs` extension rather than `.js`. I am forced to rename the file after running the `tsc` compiler. Why is it so hard to just ask the compiler to output a different extension from `.js`?

**Comment by RyanCavanaugh** [comment]
> Why is there so much resistance to just telling the compiler to output a different extension from .js?

You are already able to do this by naming your input files `.cts`

**Comment by jedwards1211** [comment]
@RyanCavanaugh I'm sure you can understand that we would already be using `.cts` if that satisfied our needs here.

To put it another way, why is there so much resistance to building in easier options for transpiling the same source to both CJS and ESM?

**Comment by ceztko** [comment]
> You are already able to do this by naming your input files .cts

@RyanCavanaugh: sometimes there's a resistance of the developer/teams to do so. I suggest again an alternative solution which I think it would be cleaner than a brutal `"targetExtension": ".cjs"` or `"targetExtension" : ".mjs"`. I was thinking something like `"ensureQualifiedExtension" : "true"` (or similarly named option) that behaves like the following:

- If we have `"module": "commonjs"` then the extension is always `.cjs`;
- If we have `"module": "es6"`, or `"es2015"`, `"es2020"`, `"es2022"`, `"esnext"` then the extension is always `.mjs`;
- If we have `"module": "node16"` or `"nodenext"` then the extension respects the rules for module detection as [documented](https://www.typescriptlang.org/docs/handbook/modules/reference.html#the-module-compiler-option) and the final extension can be either `.cjs` or `.mjs`.

**Comment by RyanCavanaugh** [comment]
> Do y'all view that as a fundamentally bad idea?

Not as much "bad" as "impossible". It's **in general** not possible to take a valid ESM program and turn it into a valid CJS program, or vice versa, because dependencies can and do present arbitrarily different shapes depending on what resolution mode you're working in.

In the cases where it is possible, it's only through a sort of syntax-only processing that any third-party tool can do; TS cannot add useful information to this process. We don't consider ourselves to be in the business of implementing or re-implementing every possible JS-related build transform that can exist (examples: TS does not minify, obfuscate, code-split, tree-shake, or bundle outputs), so we recommend using one of those tools if you're in a situation where a mechanical CJS<->ESM transform would work close enough for your scenario.

> I was thinking something like a "ensureQualifiedExtension" : "true" (or similarly named option) that behaves like the following

Why do you need `.cjs` output file extensions? Setting `"type": "commonsjs"` in `package.json` should generally have the same results.

See also https://www.typescriptlang.org/docs/handbook/modules/appendices/esm-cjs-interop.html#conclusions

**Comment by ceztko** [comment]
> Why do you need `.cjs` output file extensions? Setting `"type": "commonsjs"` in `package.json` should generally have the same results

Oh yes, I knew it already. In a project I also ended pushing a `package.json` with with a single line in the output folder:

```
{
  "type": "commonjs"
}
```

At the time I considered this solution as an acceptable trade-off for the inability of the compiler to unconditionally set a qualified extension. I still think that if the compiler could be told to emit a qualified extension it would be a more elegant solution.

**Comment by jedwards1211** [comment]
> Why do you need .cjs output file extensions? Setting "type": "commonsjs" in package.json should generally have the same results.

So, many packages output CJS and ESM to separate directories and use a `package.json` in each to control module type.

This works for importing the root of the package, but breaks down when you want to support subpath imports like `import { bar } from 'foo/bar'`.  You could have export map rules like `"./*": { "types": { "import": "./dist/esm/*.d.ts", "default": "./dist/cjs/*.d.ts" }, "import": "./dist/esm/*.js", "default": "./dist/cjs/*.js" }`, but if a user has the default TypeScript `moduleResolution` setting, TypeScript will ignore the export map and fail to find `'foo/bar'`.  In that case it can only find it if `./bar.js` exists in the package root directory.  So the only option is to build either the CJS or the ESM to the package root directory.

I mean I guess you could build CJS into the root directory and then ESM into a `dist/esm` subdirectory, since TS is only going to resolve the ESM in a mode that consults export maps anyway.  But that feels messier than outputting `.mjs` files parallel to the `.js` files (or vice versa, with `"type": "module"` and `.js`/`.cjs`).

If there are any simpler ways to support all module types and resolution modes with subpath imports, I'm all ears.

**Comment by jedwards1211** [comment]
> It's in general not possible to take a valid ESM program and turn it into a valid CJS program, or vice versa, because dependencies can and do present arbitrarily different shapes depending on what resolution mode you're working in.

I guess you wouldn't want to provide this feature unless it were possible for it to work perfectly, but I wouldn't expect it to figure out what shape the imported module would have (since that depends on the runtime environment anyway).  There are cases where I have to call an interop helper in my source code to make the transpiled ESM work in Node.  I would have to call that interop helper anyway if my package were pure ESM.
I just wish it TS would merely transform import sources to resolved paths.  Since TS-transpiled ESM importing from CJS is subject to the limitations of how Node and other tools analyze the exports of the CJS file, I would expect the same limitations if TS provided an option to transform output extensions and import paths.

**Comment by RyanCavanaugh** [comment]
> I just wish TS would merely transform import sources to resolved paths, since it would be cleaner and more foolproof than what build tools do

I don't understand this statement. One of two things is true:

 * The mapping between import source and resolved path is *trivial*, in which case all solutions are equally foolproof
 * The mapping between import source and resolved path is *nontrivial*, in which case it can't possibly be foolproof at all - just try to make sense of the output of `--traceResolution` or consider the length of documentation needed to explain module resolution at a high level. Adding an extra dimension to that matrix (what is the transform of an arbitrary path in the presence of arbitrary config) can't possibly make things easier to understand

**Comment by morganney** [comment]
@jedwards1211 all module resolutions except `classic` support package.json `imports` and `exports` fields. Maybe `node10` is an exception too, but nobody should be using those module resolutions anyway.

TS will never rewrite extensions or imports so finding a third-party tool is your best bet.

**Comment by jedwards1211** [comment]
@morganney isn't node10 still the default from `tsc --init`/omitting the field or has that changed?  It doesn't matter what people should use if the default is still widespread

**Comment by RyanCavanaugh** [comment]
> use a third party compiler (which is less of an authority on TS syntax and semantics)

Right, and what you're proposing - that TS have some notion of how to rewrite import specifiers - makes this problem worse, not better. There's a rich and excellent ecosystem of other tools that can transform TS to JS, many of which can do their jobs very correctly without needing to import/replicate TS, and making the *runtime behavior* of an output TS program dependent on TS's resolution semantics would make this tools' jobs much much harder.

**Comment by jedwards1211** [comment]
@RyanCavanaugh  I'm talking about the fact that if TS adds a new syntax in the future, third party tools will lag behind in their support of it.  Can you understand how that feels unnerving?  And it would feel more secure to only have to depend on tsc?

**Comment by morganney** [comment]
Accept that ESM and CJS are not generally compatible. For instance no live bindings or TLA in CJS. Then do this:

1. Run your compilation with tsc against a module system, say esm.
2. Dynamically alter `type` in package.json and `module`/`moduleResolution` in TS config.json to the other module system, say cjs.
3.Run another compilation with tsc against the dynamic config changes.
4. Rewrite extensions and specifiers for the output of the dual build.
5. Restore original configs.
